lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+
+		/*
+		 * Since we could be asking to modify attributes in a different
+		 * order than where they appear in the HTMl stream we need to
+		 * sort the modification replacements so we can progress
+		 * linearly and in a single pass.
+		 */
+		usort( $this->modifications->replacements, static function ( $a, $b ) {
+			if ( $a[0] === $b[0] ) {
+				return 0;
+			}
+
+			return $a[0] < $b[0] ? -1 : 1;
+		} );
+
+		$output = '';
+		$at = 0;
+		foreach ( $this->modifications->replacements as $replacement ) {
+			list( $start, $end, $text ) = $replacement;
+
+			if ( $start > $at ) {
+				$output .= substr( $document, $at, $start - $at );
+			}
+
+			$output .= $text;
+			$at = $end;
+		}
+
+		$output .= substr( $document, $at );
+		return $output;
+	}
+
+
+	public function __toString() {
+		return $this->apply();
+	}
+}
+
+
+/**
+ * Describes the search conditions for finding a given tag in an HTML document.
+ */
+class WP_Tag_Find_Descriptor {
+	/**
+	 * We're looking for an HTML tag of this name, up to the comparable
+	 * equivalence of those names (lower-cased, Unicode-normalized, etc...).
+	 * If we're looking for "any tag" then this property will be `null`.
+	 *
+	 * h1...h6 are special since they are variations of the same base tag.
+	 * To find "any heading tag" pass the special value `h`.
+	 *
+	 * @var string|null
+	 */
+	public $tag_name;
+
+
+	/**
+	 * We're looking for a tag also containing this CSS class name, up to
+	 * the comparable equivalence of those names. If we're not looking for
+	 * a class name this property will be `null`.
+	 *
+	 * @var string|null
+	 */
+	public $class_name;
+
+
+	/**
+	 * We're looking for a tag that also has some data-attribute values.
+	 * If all we care about is the presence of one of those values this
+	 * will contain the name of that attribute, up to the comparable
+	 * equivalence of that name. If we also want to constrain the value
+	 * of that attribute we will have a pair containing the name of the
+	 * attribute (to equivalence) with a predicate function. If we're
+	 * not looking for a data-attribute this property will be `null`.
+	 *
+	 * Example:
+	 * <code>
+	 *     // has a `data-wp-container` attribute
+	 *     'data-wp-container'
+	 *
+	 *     // has a `data-wp-container` attribute with the value `core/group`
+	 *     ['data-wp-block', 'core/group']
+	 *
+	 *     // has a `data-test-group` attribute with an even-valued id
+	 *     ['data-test-group', function ( $group ) { return 0 === (int) $group % 2; }]
+	 * </code>
+	 *
+	 * @var array<string|callable>
+	 */
+	public $data_attribute;
+
+
+	/**
+	 * Used to skip matches in case we expect more than one to exist.
+	 * This constraint applies after all other constraints have held.
+	 * For example, to find the third `<div>` tag containing the
+	 * `wp-block` class name set `match_offset = 3`. If not provided
+	 * the default indication is to find the first match.
+	 *
+	 * @default 1
+	 * @var int
+	 */
+	public $match_offset;
```

Comment:
```suggestion
	private $tag_name;


	/**
	 * We're looking for a tag also containing this CSS class name, up to
	 * the comparable equivalence of those names. If we're not looking for
	 * a class name this property will be `null`.
	 *
	 * @var string|null
	 */
	private $class_name;


	/**
	 * We're looking for a tag that also has some data-attribute values.
	 * If all we care about is the presence of one of those values this
	 * will contain the name of that attribute, up to the comparable
	 * equivalence of that name. If we also want to constrain the value
	 * of that attribute we will have a pair containing the name of the
	 * attribute (to equivalence) with a predicate function. If we're
	 * not looking for a data-attribute this property will be `null`.
	 *
	 * Example:
	 * <code>
	 *     // has a `data-wp-container` attribute
	 *     'data-wp-container'
	 *
	 *     // has a `data-wp-container` attribute with the value `core/group`
	 *     ['data-wp-block', 'core/group']
	 *
	 *     // has a `data-test-group` attribute with an even-valued id
	 *     ['data-test-group', function ( $group ) { return 0 === (int) $group % 2; }]
	 * </code>
	 *
	 * @var array<string|callable>
	 */
	private $data_attribute;


	/**
	 * Used to skip matches in case we expect more than one to exist.
	 * This constraint applies after all other constraints have held.
	 * For example, to find the third `<div>` tag containing the
	 * `wp-block` class name set `match_offset = 3`. If not provided
	 * the default indication is to find the first match.
	 *
	 * @default 1
	 * @var int
	 */
	private $match_offset;
```

I'd make all the properties private to limit the public API surface.